base_layer/core/src/transaction_protocol/sender.rs

```diff
@@ -97,6 +112,228 @@ impl SenderTransactionProtocol {
+        }
+    }
+
+    /// Build the sender's message for the single-round protocol (one recipient) and move to next State
+    pub fn build_single_round_message(&mut self) -> Result<SingleRoundSenderData, TPE> {
+        match &self.state {
+            SenderState::SingleRoundMessageReady(info) => {
+                let result = SingleRoundSenderData {
+                    tx_id: info.ids[0],
+                    amount: self.get_total_amount().unwrap(),
+                    public_nonce: info.public_nonce.clone(),
+                    public_excess: info.public_excess.clone(),
+                    metadata: info.metadata.clone(),
+                };
+                self.state = SenderState::CollectingSingleSignature(info.clone());
+                Ok(result)
+            },
+            _ => Err(TPE::InvalidStateError),
+        }
+    }
+
+    /// Add the signed transaction from the recipient and move to the next state
+    pub fn add_single_recipient_info(&mut self, rec: RecipientSignedTransactionData) -> Result<(), TPE> {
+        match &mut self.state {
+            SenderState::CollectingSingleSignature(info) => {
+                // Consolidate transaction info
+                info.outputs.push(rec.output);
+                // nonce is in the signature, so we'll add those together later
+                info.public_excess = &info.public_excess + &rec.public_spend_key;
+                info.public_nonce_sum = &info.public_nonce_sum + rec.partial_signature.get_public_nonce();
+                info.signatures.push(rec.partial_signature);
+                // TODO check range proof
+                self.state = SenderState::Finalizing(info.clone());
+                Ok(())
+            },
+            _ => Err(TPE::InvalidStateError),
+        }
+    }
+
+    /// Attempts to build the final transaction.
+    fn build_transaction(info: &RawTransactionInfo, features: KernelFeatures) -> Result<Transaction, TPE> {
+        let mut tx_builder = TransactionBuilder::new();
+        for i in &info.inputs {
+            tx_builder.add_input(i.clone());
+        }
+
+        for o in &info.outputs {
+            tx_builder.add_output(o.clone());
+        }
+        tx_builder.add_offset(info.offset.clone());
+        let mut s_agg = info.signatures[0].clone();
+        info.signatures.iter().skip(1).for_each(|s| s_agg = &s_agg + s);
+        let excess = CommitmentFactory::from_public_key(&info.public_excess);
+        let kernel = KernelBuilder::new()
+            .with_fee(info.metadata.fee)
+            .with_features(features)
+            .with_lock_height(info.metadata.lock_height)
+            .with_excess(&excess)
+            .with_signature(&s_agg)
+            .build()?;
+        tx_builder.with_kernel(kernel);
+        let tx = tx_builder.build()?;
+        Ok(tx)
+    }
+
+    /// Performs sanitary checks on the collected transaction pieces prior to building the final Transaction instance
```

Comment:
Do we need rubber gloves for this? Sanity?

Reply:
LOL. true